doc/rspamc.1
.\" Automatically generated by Pandoc 3.1.11.1
.\"
.TH "RSPAMC" "1" "" "Rspamd User Manual" ""
.SH NAME
\f[CR]rspamc\f[R] \- rspamd command line client
.SH SYNOPSIS
rspamc [\f[I]options\f[R]] [\f[I]command\f[R]]
[\f[I]input\-file\f[R]]\&...
.PP
rspamc \[en]help
.SH DESCRIPTION
\f[CR]rspamc\f[R] is a simple rspamd client, primarily for classifying
or learning messages.
\f[CR]rspamc\f[R] supports the following commands:
.IP \[bu] 2
Scan commands:
.RS 2
.IP \[bu] 2
\f[CR]symbols\f[R]: scan message and show symbols (default command)
.RE
.IP \[bu] 2
Control commands
.RS 2
.IP \[bu] 2
\f[CR]learn_spam\f[R]: learn message as spam
.IP \[bu] 2
\f[CR]learn_ham\f[R]: learn message as ham
.IP \[bu] 2
\f[CR]fuzzy_add\f[R]: add message to fuzzy storage (check \f[CR]\-f\f[R]
and \f[CR]\-w\f[R] options for this command)
.IP \[bu] 2
\f[CR]fuzzy_del\f[R]: delete message from fuzzy storage (check
\f[CR]\-f\f[R] option for this command)
.IP \[bu] 2
\f[CR]stat\f[R]: show rspamd statistics
.IP \[bu] 2
\f[CR]stat_reset\f[R]: show and reset rspamd statistics (useful for
graphs)
.IP \[bu] 2
\f[CR]counters\f[R]: display rspamd symbols statistics
.IP \[bu] 2
\f[CR]uptime\f[R]: show rspamd uptime
.IP \[bu] 2
\f[CR]add_symbol\f[R]: add or modify symbol settings in rspamd
.IP \[bu] 2
\f[CR]add_action\f[R]: add or modify action settings
.RE
.PP
Control commands that modify rspamd state are considered privileged and
require a password to be specified with the \f[CR]\-P\f[R] option (see
\f[B]OPTIONS\f[R], below, for details).
.PD 0
.P
.PD
This depends on a controller\[cq]s settings and is discussed in the
\f[CR]rspamd\-workers\f[R] page (see \f[B]SEE ALSO\f[R], below, for
details).
.PP
\f[CR]Input files\f[R] may be either regular file(s) or a directory to
scan.
If no files are specified \f[CR]rspamc\f[R] reads from the standard
input.
Controller commands usually do not accept any input, however learn* and
fuzzy* commands requires input.
.SH OPTIONS
.TP
\-h \f[I]host[:port]\f[R], \-\-connect=\f[I]host[:port]\f[R]
Specify host and port
.TP
\-P \f[I]password\f[R], \-\-password=\f[I]password\f[R]
Specify control password. Can be an absolute or relative path, in which
case the password will be read from that file.
.TP
\-c \f[I]name\f[R], \-\-classifier=\f[I]name\f[R]
Classifier to learn spam or ham (bayes is used by default)
.TP
\-w \f[I]weight\f[R], \-\-weight=\f[I]weight\f[R]
Weight for fuzzy operations
.TP
\-f \f[I]number\f[R], \-\-flag=\f[I]number\f[R]
Flag for fuzzy operations
.TP
\-p, \-\-pass
Pass all filters
.TP
\-v, \-\-verbose
More verbose output
.TP
\-i \f[I]ip address\f[R], \-\-ip=\f[I]ip address\f[R]
Emulate that message was received from specified ip address
.TP
\-u \f[I]username\f[R], \-\-user=\f[I]username\f[R]
Emulate that message was received from specified authenticated user
.TP
\-d \f[I]user\[at]domain\f[R], \-\-deliver=\f[I]user\[at]domain\f[R]
Emulate that message was delivered to specified user (for
LDA/statistics)
.TP
\-F \f[I]user\[at]domain\f[R], \-\-from=\f[I]user\[at]domain\f[R]
Emulate that message has specified SMTP FROM address
.TP
\-r \f[I]user\[at]domain\f[R], \-\-rcpt=\f[I]user\[at]domain\f[R]
Emulate that message has specified SMTP RCPT address
.TP
\-\-helo=\f[I]helo_string\f[R]
Imitate SMTP HELO passing from MTA
.TP
\-\-hostname=\f[I]hostname\f[R]
Imitate hostname passing from MTA (rspamd assumes that it is verified by
MTA)
.TP
\-t \f[I]seconds\f[R], \-\-timeout=\f[I]seconds\f[R]
Timeout for waiting for a reply (can be floating point number,
e.g.\ 0.1)
.TP
\-b \f[I]host:port\f[R], \-\-bind=\f[I]host:port\f[R]
Bind to specified ip address
.TP
\-R, \-\-human
Output human readable report.
The first line of the output contains the message score and three
threshold scores, in this format:
.IP
.EX
    score/greylist/addheader/reject,action=N:ACTION,spam=0|1,skipped=0|1
.EE
.TP
\-j, \-\-json
Output formatted JSON
.TP
\-\-ucl
Output UCL
.TP
\-\-raw
Output raw data received from rspamd (compacted JSON)
.TP
\-\-headers
Output HTTP headers from a reply
.TP
\-\-extended\-urls
Output URLs in an extended format, showing full URL, host and the part
of host that was used by surbl module (if enabled).
.TP
\-n \f[I]parallel_count\f[R], \-\-max\-requests=\f[I]parallel_count\f[R]
Maximum number of requests to rspamd executed in parallel (8 by default)
.TP
\-e \f[I]command\f[R], \-\-execute=\f[I]command\f[R]
Execute the specified command with either mime output (if
\f[CR]mime\f[R] option is also specified) or formatted rspamd output
.TP
\-\-mime
Output the full mime message instead of scanning results only
.TP
\-\-header=\f[I]header\f[R]
Add custom HTTP header for a request.
You may specify header in format \f[CR]name=value\f[R] or just
\f[CR]name\f[R] for an empty header.
This option can be repeated multiple times.
.TP
\-\-sort=\f[I]type\f[R]
Sort output according to a specific field.
For \f[CR]counters\f[R] command the allowed values for this key are
\f[CR]name\f[R], \f[CR]weight\f[R], \f[CR]frequency\f[R] and
\f[CR]hits\f[R].
Appending \f[CR]:asc\f[R] to any of these types inverts sorting order.
.TP
\-\-commands
List available commands
.SH RETURN VALUE
On exit \f[CR]rspamc\f[R] returns \f[CR]0\f[R] if operation was
successful and an error code otherwise.
.SH EXAMPLES
Check stdin:
.IP
.EX
rspamc < some_file
.EE
.PP
Check files:
.IP
.EX
rspamc symbols file1 file2 file3
.EE
.PP
Learn files:
.IP
.EX
rspamc \-P pass learn_spam file1 file2 file3
.EE
.PP
Add fuzzy hash to set 2:
.IP
.EX
rspamc \-P pass \-f 2 \-w 10 fuzzy_add file1 file2
.EE
.PP
Delete fuzzy hash from other server:
.IP
.EX
rspamc \-P pass \-h hostname:11334 \-f 2 fuzzy_del file1 file2
.EE
.PP
Get statistics:
.IP
.EX
rspamc stat
.EE
.PP
Get uptime:
.IP
.EX
rspamc uptime
.EE
.PP
Add custom rule\[cq]s weight:
.IP
.EX
rspamc add_symbol test 1.5
.EE
.PP
Add custom action\[cq]s weight:
.IP
.EX
rspamc add_action reject 7.1
.EE
.SH SEE ALSO
Rspamd documentation and source code may be downloaded from \c
.UR https://rspamd.com/
.UE \c
\&.


ChangeLog
3.11.1: 08 Mar 2025
  * [Feature] Add 'noop' redis backend for scripts running
  * [Feature] Add Redis caching framework
  * [Feature] Add UTF8 `sub` and `len` variants for rspamd_text
  * [Feature] Allow adding timers to task (respecting symbols)
  * [Feature] Allow additional categories to be defined in GPT
  * [Feature] Allow fine-grained control on keys permissions
  * [Feature] Allow individual `read_only` flag per key
  * [Feature] Allow multiple lua scripts for fuzzy storage
  * [Feature] Allow to add periodic functions in Lua API
  * [Feature] Allow to disable rbls from map
  * [Feature] Allow to hash any Lua types
  * [Feature] Allow to store shingles as opaque Lua data
  * [Feature] Cache LLM replies
  * [Feature] GPT: Add ollama support
  * [Feature] GPT: Support reason adding
  * [Feature] Improve prompt and use plaintext instead of JSON
  * [Feature] Lua_task: Allow to load data into the existing task
  * [Feature] More additions
  * [Feature] Pass shingles to Lua scripts
  * [Feature] Preliminary implementation of LLM based anonymizing
  * [Feature] Support LLM models consensus
  * [Feature] Try to check maps earlier if their expires is too long
  * [Feature] Use debug module name from caller in lua_cache
  * [Fix] Add timer update before timer setting
  * [Fix] Allow to work with no ratelimits
  * [Fix] Always create ratelimit buckets
  * [Fix] Avoid collision hacks in mempool variables hash
  * [Fix] Expire neural ham and spam sets so they not hangup in redis indefinitely
  * [Fix] Fix crash on FreeBSD when Rspamd is built without hyperscan
  * [Fix] Make table digests consistent
  * [Fix] RBL: fix use of `content_urls` and `images` inside `checks`
  * [Fix] Verify key type to match DKIM signature type
  * [Fix] connIP is not correctly added to request
  * [Fix] properly close multipart/related boundary when adding text footer
  * [Rework] GPT: Use cache framework

3.11.0: 16 Dec 2024
  * [Conf] Add lua.local.d folder
  * [Conf] Add more ways to extend Rspamd configuration
  * [Conf] Allow to install examples conditionally
  * [Conf] Fix comments in elastic.conf
  * [Conf] Install examples
  * [Conf] Use full hostnames for SURBL
  * [CritFix] Allow to reupload scripts when Redis is restarted
  * [CritFix] Fix ARC-Seal signing
  * [CritFix] Properly check dynamic keys in fuzzy storage
  * [Feature] Add LRU cache for last filled ratelimit buckets
  * [Feature] Add include/exclude logic for headers
  * [Feature] Add ratelimit util to manage last filled ratelimit buckets
  * [Feature] Dump current ratelimits for fuzzy keys
  * [Feature] GPT: Try harder to find JSON in NN reply
  * [Feature] Improve address rotation algorithm
  * [Feature] Show stats in rspamadm
  * [Feature] Unify displayed part selection
  * [Fix] Avoid null-bytes in Log-Tag header value.
  * [Fix] Another story about char sign
  * [Fix] Check DNS limits when resolving MX/PTR records
  * [Fix] Dmarc: Properly encode structured headers
  * [Fix] Do not abort when OpenSSL is broken, report that to a user
  * [Fix] Fix dealing with happy eyeballs in SPF
  * [Fix] Fix memory leak in `lua_new_text` invocations
  * [Fix] Fix usage of the logging function
  * [Fix] Load "default" provider as well as "legacy"
  * [Fix] Milter headers: fold before encoding
  * [Fix] More fixes to rfc2047 encoding
  * [Fix] More things to fix
  * [Fix] Phishing: Fix many cases of FP for the same domains
  * [Fix] Remove USPOOF_SINGLE_SCRIPT_CONFUSABLE from the checks
  * [Fix] Restore propagation of spam/ham learns
  * [Fix] Sigh, another fix
  * [Fix] Some fixes to `rspamadm ratelimit`
  * [Fix] Some more fixes
  * [Fix] Use `dkim_signing` for `sign_headers` option
  * [Fix] Use correct type for keylen in lua_ucl_newindex
  * [Fix] Use cumulative timeout when dealing with TCP connections
  * [Fix] Use unsigned char for ragel machines
  * [Project] Add routine to strip attachments
  * [Project] Add tool to rspamadm
  * [Project] Initial stuff to implement messages anonymization
  * [Project] Keep Changelog updated
  * [Project] Various fixes and `rspamadm mime strip` command
  * [Rework] Allow `Content-Encoding` standard header for zstd compression
  * [Rework] Breaking: Actualize elastic module, support Elastic 8 & OpenSearch 2, add index policy with logs retention and many more
  * [Rework] Multimap should use only distinct text parts for content matching
  * [Rework] Replace fastutf with simdutf
  * [Rework] Rewrite rfc2047 encoding as it was totally broken
  * [Rework] Update hiredis to 1.2.0

3.10.2: 21 Oct 2024
  * [CritFix] Fix ARC-Seal signing
  * [Fix] add EOF to openmetrics response in proxy and server

3.10.1: 16 Oct 2024
  * [Feature] Update effective_tld_names.dat by @wdhdev in #5176
  * [Fix] Use correct type for keylen in lua_ucl_newindex by @arkamar in #5169
  * [Fix] Avoid null-bytes in Log-Tag header value by @smarsching in #5179
  * [Fix] Do not abort when OpenSSL is broken, report that to a user by @vstakhov in #5188
  * [Fix] Update hiredis library removing all hacks by @vstakhov in #5167
  * [Fix] Remove proxy from url_redirector.conf as it not the option by @dragoangel in #5164
  * [Fix] Some build fixes by @vstakhov in #5189
  * [Fix] Some more fixes by @vstakhov in #5190

3.10.0: 30 Sep 2024
  * [Conf] Add SenderScore RPBL return codes
  * [Conf] Add SenderScore Reputationlist RBL
  * [Conf] Increase scores for strange things in the archives
  * [CritFix] The max size for signing key is actually 65 bytes for p256
  * [Feature] Add rspamadm secretbox command
  * [Feature] Add rspamd_cxx_unit_cryptobox for unit testing
  * [Feature] Add support for OpenSSL 3.0
  * [Feature] Add support for OpenSSL 3.0 for DKIM
  * [Feature] Add tests for rspamd_cryptobox
  * [Feature] Add tooling to encrypt strings in Lua
  * [Feature] Allow differen modes for fuzzy rules
  * [Feature] Allow to set negative group score limit via `min_score`
  * [Feature] DMARC: Implement reporting.only_domains setting

rspamc.pl
#!/usr/bin/perl -w

# Simple script that read message from STDIN and test it on rspamd server
# using specified command.
#
# Usage: rspamc.pl [-c conf_file] [command] [-s statfile]
#
# By default rspamc.pl would read ./rspamd.conf and default command is SYMBOLS

use Socket qw(:DEFAULT :crlf);

my %cfg = (
    'conf_file' => './rspamd.conf',
    'command'   => 'SYMBOLS',
    'host'      => 'localhost',
    'port'      => '11333',
    'is_unix'   =>  0,
    'password'  =>  '',
    'control'   =>  0,
    'statfile'  =>  '',
);


sub usage {
    return "Usage: rspamc.pl [-c conf_file] [-s statfile] [command]";
}

# Load rspamd config params
sub parse_config {
    my ($is_ctrl) = @_;

    open CONF, "< $cfg{'conf_file'}" or die "config file $cfg{'conf_file'} cannot be opened";

    my $ctrl = 0;
    while (<CONF>) {
        if ($_ =~ /control\s*{/i) {
            $ctrl = 1;
        }
        if ($ctrl && $_ =~ /}/) {
            $ctrl = 0;
        }
        if (((!$is_ctrl && !$ctrl) || ($ctrl && $is_ctrl))
                && $_ =~ /^\s*bind_socket\s*=\s*((([^:]+):(\d+))|(\/\S*))/i) {
            if ($3 && $4) {
                $cfg{'host'} = $3;
                $cfg{'port'} = $4;
                $cfg{'is_unix'} = 0;
            }
            else {
                $cfg{'host'} = $5;
                $cfg{'is_unix'} = 1;
            }
        }
        if ($ctrl && $is_ctrl && $_ =~ /^\s*password\s*=\s*"(\S+)"/) {
            $cfg{'password'} = $1;
        }
    }

    close CONF;

}

sub connect_socket {
    my $sock;

    if ($cfg{'is_unix'}) {
        socket ($sock, PF_UNIX, SOCK_STREAM, 0) or die "cannot create unix socket";
        my $sun = sockaddr_un($cfg{'host'});
        connect ($sock, $sun) or die "cannot connect to socket $cfg{'host'}";
    }
    else {
        my $proto = getprotobyname('tcp');
        my $sin;
        socket ($sock, PF_INET, SOCK_STREAM, $proto) or die "cannot create tcp socket";
        if (inet_aton ($cfg{'host'})) {
            $sin = sockaddr_in ($cfg{'port'}, inet_aton($cfg{'host'}));
        }
        else {
            my $addr = gethostbyname($cfg{'host'});
            if (!$addr) {
                die "cannot resolve $cfg{'host'}";
            }
            $sin = sockaddr_in ($cfg{'port'}, $addr);
        }
        
        connect ($sock, $sin) or die "cannot connect to socket $cfg{'host'}:$cfg{'port'}";
    }
    return $sock;
}

# Currently just read stdin for user's message and pass it to rspamd
sub do_rspamc_command {
    my ($sock) = @_;

    my $input;
    while (defined (my $line = <>)) {
        $input .= $line;
    }

    print "Sending ". length ($input) ." bytes...\n";

    syswrite $sock, "$cfg{'command'} RSPAMC/1.0 $CRLF";
    syswrite $sock, "Content-Length: " . length ($input) . $CRLF . $CRLF;
    syswrite $sock, $input;
    syswrite $sock, $CRLF;
    while (defined (my $line = <$sock>)) {
        print $line;
    }
}

sub do_ctrl_auth {
    my ($sock) = @_;

    syswrite $sock, "password $cfg{'password'}" . $CRLF;
    if (defined (my $reply = <$sock>)) {
        my $end = <$sock>;
        if ($reply =~ /^password accepted/) {
            return 1;
        }
    }

    return 0;
}

sub do_control_command {
    my ($sock) = @_;

    # Read greeting first
    if (defined (my $greeting = <$sock>)) {
        if ($greeting !~ /^Rspamd version/) {
            die "not rspamd greeting line $greeting";
        }
    }
    if ($cfg{'command'} =~ /^learn$/i) {
        my $input;
        die "statfile is not specified to learn command" if !$cfg{'statfile'};

        while (defined (my $line = <>)) {
            $input .= $line;
        }
        
        if (do_ctrl_auth ($sock)) {
            my $len = length ($input);
            print "Sending $len bytes...\n";
            syswrite $sock, "learn $cfg{'statfile'} $len" . $CRLF;
            syswrite $sock, $input . $CRLF;
            if (defined (my $reply = <$sock>)) {
                if ($reply =~ /^learn ok/) {
                    print "Learn succeed\n";
                }
                else {
                    print "Learn failed\n";
                }
            }
        }
        else {
            print "Authentication failed\n";
        }
    }
    elsif ($cfg{'command'} =~ /(reload|shutdown)/i) {
        if (do_ctrl_auth ($sock)) {
            syswrite $sock, $cfg{'command'} . $CRLF;
            while (defined (my $line = <$sock>)) {
                last if $line =~ /^END/;
                print $line;
            }
        }
        else {
            print "Authentication failed\n";
        }
    }
    else {
        syswrite $sock, $cfg{'command'} . $CRLF;
        while (defined (my $line = <$sock>)) {
            last if $line =~ /^END/;
            print $line;
        }
    }
}

while (my $param = shift) {
    if ($param eq '-c') {
        my $value = shift;
        if ($value) {
            if (-r $value) {
                $cfg{'conf_file'} = $value;
            }
            else {
                die "config file $value is not readable";
            }
        }
        else {
            die usage();
        }
    }
    elsif ($param eq '-s') {
        my $value = shift;
        if ($value) {
            $cfg{'statfile'} = $value;
        }
        else {
            die usage();
        }
    }
    elsif ($param =~ /(SYMBOLS|SCAN|PROCESS|CHECK|REPORT_IFSPAM|REPORT)/i) {
        $cfg{'command'} = $1;
        $cfg{'control'} = 0;
    }
    elsif ($param =~ /(STAT|LEARN|SHUTDOWN|RELOAD|UPTIME)/i) {
        $cfg{'command'} = $1;
        $cfg{'control'} = 1;
    }
    else {
        die usage();
    }
}

parse_config ($cfg{'control'});
my $sock = connect_socket ();

if ($cfg{'control'}) {
    do_control_command ($sock);
}
else {
    do_rspamc_command ($sock);
}

close ($sock);
